pkg/grpc/jwt_authenticator_test.go

```diff
@@ -0,0 +1,131 @@
+package grpc_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/buildbarn/bb-storage/internal/mock"
+	bb_grpc "github.com/buildbarn/bb-storage/pkg/grpc"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/jwt"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+func TestJWTAuthenticator(t *testing.T) {
+	ctrl, ctx := gomock.WithContext(context.Background(), t)
+	defer ctrl.Finish()
+	clock := mock.NewMockClock(ctrl)
+
+	symmetricKey := []byte("0123456789ABCDEF")
+
+	jwtKey := bb_grpc.JWTKeyConfig{
+		Key: symmetricKey,
+	}
+
+	signer := mustMakeSigner(jose.HS256, symmetricKey)
+
+	authenticator := bb_grpc.NewJWTAuthenticator(jwtKey, clock)
+
+	t.Run("NoGRPC", func(t *testing.T) {
+		// Authenticator is used outside of gRPC, meaning it cannot
+		// extract request metadata.
+		require.Equal(
+			t,
+			status.Error(codes.Unauthenticated, "Connection was not established using gRPC"),
+			authenticator.Authenticate(ctx))
+	})
+
+	t.Run("NoAuthorizationMetadata", func(t *testing.T) {
+		// Should deny authentication if no `authorization` header is present.
+		md := metadata.MD{}
+		require.Equal(
+			t,
+			status.Error(codes.Unauthenticated, "Missing authorization header"),
+			authenticator.Authenticate(metadata.NewIncomingContext(ctx, md)),
+		)
+	})
+
+	t.Run("HasAuthorizationMetadataKeyButNoValues", func(t *testing.T) {
+		// Should deny authentication if `authorization` header is present but has no values.
+		md := metadata.MD{
+			"authorization": nil,
+		}
+		require.Equal(
+			t,
+			status.Error(codes.Unauthenticated, "Missing authorization header"),
+			authenticator.Authenticate(metadata.NewIncomingContext(ctx, md)),
+		)
+	})
+
+	t.Run("HasAuthorizationMetadataKeyButMultipleValues", func(t *testing.T) {
+		// Should deny authentication if `authorization` header is present and has multiple values.
+		md := metadata.Pairs("authorization", "hello", "authorization", "world")
+		require.Equal(
+			t,
+			status.Error(codes.Unauthenticated, "Multiple authorization headers are not supported"),
+			authenticator.Authenticate(metadata.NewIncomingContext(ctx, md)),
+		)
+	})
+
+	t.Run("ParsesAndValidateValidJWS", func(t *testing.T) {
+		// Should parse and validate a valid JWS.
+		clock.EXPECT().Now().Return(time.Unix(1600000000, 0))
+		tok, err := jwt.Signed(signer).
```

Comment:
Please [don't put logic in tests](https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html). Instead of generating a token here, please use a hardcoded one that corresponds with the symmetric key. If we were to use a version of go-jose that generates & validates non-standard tokens, this test wouldn't catch it.

Reply:
The key is already hard-coded in the test. See line 25:

```
symmetricKey := []byte("0123456789ABCDEF")
```

`signer` is just the go-jose wrapper around the key.

Reply:
Exactly. But I'm suggesting that the token against which you validate is also hardcoded.